src/blog/2023/02/outage-report-2023-01-27.md

```diff
@@ -0,0 +1,124 @@
+
+## Technical Details
+
+When running FlowForge within a Kubernetes environment, each project creates a
+new Ingress Object configuration to tell the platform how to route HTTP traffic
+to that project.
+
+Our FlowForge Cloud deployment runs within Amazon Elastic Kubernetes Service (EKS)
+and uses the Application Load Balancer (ALB) service as its ingress controller.
+
+When the FlowForge platform creates the new Ingress Object configuration, EKS passes
+that to ALB to generate the necessary configuration, which, given the configuration we
+were using, created both a Target Group and Rule object.
+
+With a default of limit of 100 Target Groups and Rules, that meant we had a technical
+limit of 100 Node-RED projects within the FlowForge Cloud environment. Increasing
+the Rule limit did not solve the problem as the Target Group limit still applied.
+
+Our initial mitigation was to delete any Node-RED projects we had created for
+our own internal testing. We also identified that we could safely delete any rules
+for suspended projects. A suspended project is one that is not actively running
+in the platform. The code that resumes a suspended project would recreate any ingress
+objects needed - so deleting the rule whilst suspended would not have any impact
+on the project.
+
+This gave us a small amount of headroom on the platform which crucially meant we
+had time to develop a longer term solution.
+
+## Resolution
+
+There were two possible routes we could take to resolve this issue:
+
+ - Investigate how to reuse existing Target Groups rather than create one for each project
+
+   Our intial research was inconclusive on how to achieve this. The AWS docs weren't
+   clear enough to give us a definitive answer we felt comfortable to invest our time
+   in.
+
+ - Move away from ALB in favour of nginx to provide our ingress load balancing.
+
+   This was always our long term strategy as it was a prerequisite to FlowForge
+   features we have in the roadmap such as providing custom domains to projects.
+   However it was potentially a large piece of work with a complicated migration
+   for the existing environment.
+
+Given it fitted with our longer-term strategic goals, we decided to move ahead
+with replacing ALB with nginx.
+
+After some initial development work and experimentation, we felt comfortable that
+the migration was not as complicated as initially feared. We would have to manually
+copy the existing ALB rules over - something that could be scripted. Once we had
+nginx deployed we could push a small code change to FlowForge to use it rather than
+ALB, and also switch over the DNS entry to point at nginx.
+
+Following a successful run through in our testing/staging environment, we decided
+to move ahead updating the production environment.
+
+This change was applied today, whilst we closely monitored the system to ensure
+no further disruption occurred.
+
+_**TODO conclusion once we know how it went**_
```

Comment:
We need to deploy X instances today so we verified we can move beyond 100 today I think